## **Урок 3. Описательная статистика. Качественные и количественные характеристики популяции. Графическое представление данных**

#### Задача 1. Даны значения зарплат из выборки выпускников: 
#### 100, 80, 75, 77, 89, 33, 45, 25, 65, 17, 30, 24, 57, 55, 70, 75, 65, 84, 90, 150. 
#### Посчитать (желательно без использования статистических методов наподобие std, var, mean) среднее арифметическое, среднее квадратичное отклонение, смещенную и несмещенную оценки дисперсий для данной выборки.

In [84]:
import numpy as np
from math import factorial

In [85]:
salaries = np.array([100, 80, 75, 77, 89, 33, 45, 25, 65, 17, 30, 24, 57, 55, 70, 75, 65, 84, 90, 150])
salaries_len = len(salaries)
salaries_len

20

In [86]:
#среднее арифметическое
arithmetic_mean = sum(salaries)/salaries_len
arithmetic_mean 

65.3

In [87]:
#проверка
arithmetic_mean_func = np.mean(salaries)
arithmetic_mean_func

65.3

Среднеквадратичное отклонение (смещенное):
$$\sigma = \sqrt{\frac{\sum\limits_{i=1}^{n} (x_i - \overline{x})^2}{n}}$$

In [88]:
salaries_std_array = np.array([(sal - arithmetic_mean) ** 2 for sal in salaries])
salaries_std_array

array([1.20409e+03, 2.16090e+02, 9.40900e+01, 1.36890e+02, 5.61690e+02,
       1.04329e+03, 4.12090e+02, 1.62409e+03, 9.00000e-02, 2.33289e+03,
       1.24609e+03, 1.70569e+03, 6.88900e+01, 1.06090e+02, 2.20900e+01,
       9.40900e+01, 9.00000e-02, 3.49690e+02, 6.10090e+02, 7.17409e+03])

In [89]:
#среднее квадратичное отклонение (смещенное)
salaries_std = np.sqrt((salaries_std_array).sum() / salaries_len)
salaries_std

30.823854398825596

In [90]:
#среднее квадратичное отклонение (смещенное) - проверка
sal_std_func = np.std(salaries)
sal_std_func

30.823854398825596

In [91]:
#среднее квадратичное отклонение (несмещенное)
salaries_std_non_shifted = np.sqrt((salaries_std_array).sum() / (salaries_len - 1))
salaries_std_non_shifted

31.624607341019814

In [92]:
#среднее квадратичное отклонение (смещенное) - проверка
sal_std_func_non_shifted = np.std(salaries, ddof=1)
sal_std_func_non_shifted

31.624607341019814

In [93]:
#смещенная дисперсия 
shifted_dispersion = salaries_std ** 2
shifted_dispersion

950.11

In [94]:
#смещенная дисперсия - проверка
shifted_dispersion_func = np.var(salaries, ddof=0)
shifted_dispersion_func

950.11

In [95]:
#несмещенная дисперсия 
non_shifted_dispersion = salaries_std_non_shifted ** 2
non_shifted_dispersion

1000.1157894736843

In [96]:
#несмещенная дисперсия - проверка
non_shifted_dispersion_func = np.var(salaries, ddof=1)
non_shifted_dispersion_func

1000.1157894736842

#### Задача 2. В первом ящике находится 8 мячей, из которых 5 - белые. 
#### Во втором ящике - 12 мячей, из которых 5 белых. 
#### Из первого ящика вытаскивают случайным образом два мяча, из второго - 4. 
#### Какова вероятность того, что 3 мяча белые?

In [97]:
#Всего три альтернативных варианта:
#1. Из первой 2 белых, из второй 1 белый, 3 черных
#2. Из первой 1 белый, 1 черный, из второй 2 белых, 2 черных
#3. Из первой 0 белых, 2 черных, из второй 3 белых, 1 черный
# Вероятности всех трех варинатов складываются по формуле полной вероятности.

def combinations(n, k):
    return int(factorial(n) / (factorial(k) * factorial(n - k)))

In [98]:
#1 Вариант
#1. Из первой 2 белых, из второй 1 белый, 3 черных

P1_var_1 = combinations(5, 2) / combinations(8, 2)

P2_var_1 = combinations(5, 1) * combinations(7, 3) / combinations(12, 4)

P_var_1 = P1_var_1 * P2_var_1

P_var_1

0.12626262626262627

In [99]:
#2 Вариант
#2. Из первой 1 белый, 1 черный, из второй 2 белых, 2 черных

P1_var_2 = combinations(5, 1) * combinations(3, 1) / combinations(8, 2)

P2_var_2 = combinations(5, 2) * combinations(7, 2) / combinations(12, 4)

P_var_2 = P1_var_2 * P2_var_2

P_var_2

0.22727272727272727

In [100]:
#3 Вариант
#3. Из первой 0 белых, 2 черных, из второй 3 белых, 1 черный

P1_var_3 = combinations(3, 2) / combinations(8, 2)

P2_var_3 = combinations(5, 3) * combinations(7, 1) / combinations(12, 4)

P_var_3 = P1_var_3 * P2_var_3

P_var_3

0.01515151515151515

In [101]:
P_task_2 = P_var_1 + P_var_2 + P_var_3
P_task_2

0.3686868686868687

#### Ответ: вероятность того, что 3 мяча белые - 0.3686868686868687 или 36,9%

#### Задача 3.  На соревновании по биатлону один из трех спортсменов стреляет и попадает в мишень. 
#### Вероятность попадания для первого спортсмена равна 0.9, для второго — 0.8, для третьего — 0.6. 
#### Найти вероятность того, что выстрел произведен:
#### a). первым спортсменом б). вторым спортсменом в). третьим спортсменом.

Так как событие - попадание стрелка в мишень уже произошло, будем использовать формулу Байеса:

$$P(B\:|\:A) = \frac{P(B) \cdot P(A\:|\:B)}{P(A)}$$

A - попадание в мишень

P(B1) = 1/3 - вероятность, что стрелять будет первый спортсмен

P(B2) = 1/3 - вероятность, что стрелять будет второй спортсмен

P(B3) = 1/3 - вероятность, что стрелять будет третий спортсмен

P(A|B1) = 0,9 - вероятность попадания у первого спортсмена

P(A|B2) = 0,8 - вероятность попадания у второго спортсмена

P(A|B3) = 0,6 - вероятность попадания у третьего спортсмена


In [102]:
# общая вероятность попадания
P_A = (1/3) * 0.9 + (1/3) * 0.8 + (1/3) * 0.6
P_A

0.7666666666666666

In [103]:
# a. Найти вероятность того, что выстрел произведен первым спортсменом
P_B1_A  = ((1/3) * 0.9) / P_A
P_B1_A

0.391304347826087

In [104]:
# б. Найти вероятность того, что выстрел произведен вторым спортсменом
P_B2_A  = ((1/3) * 0.8) / P_A
P_B2_A

0.3478260869565218

In [105]:
# в. Найти вероятность того, что выстрел произведен третьим спортсменом
P_B3_A  = ((1/3) * 0.6) / P_A
P_B3_A

0.2608695652173913

#### Ответ: вероятность того, что выстрел произведен: 
#### a). первым спортсменом  - 0.391304347826087 или 39%,
#### б). вторым спортсменом - 0.3478260869565218 или 34,8%,
#### в). третьим спортсменом - 0.2608695652173913 или 26%. 

#### Задача 4.  В университет на факультеты A и B поступило равное количество студентов, 
#### а на факультет C студентов поступило столько же, сколько на A и B вместе. 
#### Вероятность того, что студент факультета A сдаст первую сессию, равна 0.8. 
#### Для студента факультета B эта вероятность равна 0.7, а для студента факультета C - 0.9. 
#### Студент сдал первую сессию. Какова вероятность, что он учится:
#### a). на факультете A б). на факультете B в). на факультете C?

Так как событие - сдача сессии студентом уже произошло, будем использовать формулу Байеса:

$$P(B\:|\:A) = \frac{P(B) \cdot P(A\:|\:B)}{P(A)}$$

A - сдача сессии студентом

In [106]:
P_B1 = 1/4 #вероятность, что студент учится на факультете A

P_B2 = 1/4 #вероятность, что студент учится на факультете B

P_B3 = 1/2 #вероятность, что студент учится на факультете C

P_A_B1 = 0.8 #вероятность сдачи сессии студента факультета A

P_A_B2 = 0.7 #вероятность сдачи сессии студента факультета B

P_A_B3 = 0.9 #вероятность сдачи сессии студента факультета C

In [107]:
# общая вероятность сдачи сессии для студентов всех факультетов

P_A = P_B1 * P_A_B1 + P_B2 * P_A_B2 + P_B3 * P_A_B3

P_A

0.825

In [108]:
#a).какая вероятность, что студент учится на на факультете A 

P_B1_A = (P_B1 * P_A_B1) / P_A

P_B1_A

0.24242424242424246

In [109]:
#б).какая вероятность, что студент учится на на факультете B

P_B2_A = (P_B2 * P_A_B2) / P_A

P_B2_A

0.21212121212121213

In [110]:
#в).какая вероятность, что студент учится на на факультете C

P_B3_A = (P_B3 * P_A_B3) / P_A

P_B3_A

0.5454545454545455

#### Ответ: вероятность того, что студент учится:
#### a). на факультете A - 0.24242424242424246 или 24,2%,
#### б). на факультете B - 0.21212121212121213 или 21,2%,
#### в). на факультете C- 0.5454545454545455 или 54,5%.

#### Задача 5.  Устройство состоит из трех деталей.
#### Для первой детали вероятность выйти из строя в первый месяц равна 0.1, 
#### для второй - 0.2, для третьей - 0.25. 
#### Какова вероятность того, что в первый месяц выйдут из строя: 
#### а). все детали 
#### б). только две детали 
#### в). хотя бы одна деталь 
#### г). от одной до двух деталей?

In [111]:
#а).вероятность, что из строя выйдут все детали 

P_all = 0.1 * 0.2 * 0.25
P_all

0.005000000000000001

In [112]:
#б). только две детали 
P_B1 = 0.1
P_B2 = 0.2
P_B3 = 0.25

#1 вариант (первая деталь цела, остальные сломались)
var_1 = (1 - P_B1) * P_B2 * P_B3 
var_1 

0.045000000000000005

In [113]:
#2 вариант (вторая деталь цела, остальные сломались)
var_2 = (1 - P_B2) * P_B1 * P_B3 
var_2

0.020000000000000004

In [114]:
#3 вариант (третья деталь цела, остальные сломались)
var_3 = (1 - P_B3) * P_B1 * P_B2
var_3

0.015000000000000003

In [115]:
P_2_pieces = var_1 + var_2 + var_3

P_2_pieces

0.08

In [116]:
#в). хотя бы одна деталь 
# 1 - вероятность, что ни одна деталь не сломалась.
P_none =  (1 - P_B3) * (1 - P_B1) * (1 - P_B2) 

P_at_least_one = 1 - P_none
P_at_least_one

0.45999999999999996

In [117]:
# г). от одной до двух деталей?
# или одна или две
# найдем вероятность, что сломается одна деталь. 

#1 вариант (первая сломалась, остальные целы)
var_1_1 = P_B1 * (1 - P_B2) * (1 - P_B3) 
var_1_1 

0.06000000000000001

In [118]:
#2 вариант (вторая сломалась, остальные целы)
var_2_1 = P_B2 * (1 - P_B1) * (1 - P_B3) 
var_2_1 

0.135

In [119]:
#3 вариант (третья сломалась, остальные целы)
var_3_1 = P_B3 * (1 - P_B1) * (1 - P_B2) 
var_3_1 

0.18000000000000002

In [120]:
P_1_piece = var_1_1 + var_2_1 + var_3_1

P_1_piece

0.375

In [121]:
P_1_or_2 = P_1_piece + P_2_pieces
P_1_or_2

0.455

#### Ответ: вероятность того, что в первый месяц выйдут из строя: 
#### а). все детали - 0.005000000000000001 или 0,5%,
#### б). только две детали - 0.08 или 8%,
#### в). хотя бы одна деталь - 0.45999999999999996 или 46%,
#### г). от одной до двух деталей - 0.455 или 45,5%. 